frontend_tests/node_tests/compose.js

```diff
@@ -317,6 +320,162 @@ people.add(bob);
+          sender_id: 101,
+          queue_id: undefined,
+          stream: '',
+          subject: '',
+          to: '["alice@example.com"]',
+          reply_to: 'alice@example.com',
+          private_message_recipient: 'alice@example.com',
+          to_user_ids: '31',
+          local_id: 1,
+        };
+        assert.equal(payload.url, '/json/messages');
+        assert.equal(_.keys(payload.data).length, 11);
+        assert.deepEqual(payload.data, single_msg);
+        payload.data.id = local_id_counter_stub;
+        payload.success(payload.data);
+        send_msg_ajax_post_called = send_msg_ajax_post_called + 1;
+    };
+    var get_events_running_called = 0;
+    var server_events_triggered = 0;
+    global.server_events = {
+        restart_get_events: function () {
+            server_events_triggered = server_events_triggered + 1;
+        },
+        assert_get_events_running: function () {
+            get_events_running_called = get_events_running_called + 1;
+        },
+    };
+    var reify_message_id_checked = 0;
+    echo.reify_message_id = function (local_id, message_id) {
+        assert.equal(typeof(local_id), 'number');
+        assert.equal(typeof(message_id), 'number');
+        reify_message_id_checked = reify_message_id_checked + 1;
+    };
+    global.patch_builtin('setTimeout', function (func) {
+        func();
+    });
+    compose.send_times_data = {};
+    $("#new_message_content").val('[foobar](https://foo.com/user_uploads/123456)');
+    $("#new_message_content").blur();
+    $("#send-status").show();
+    $("#compose-send-button").attr('disabled', 'disabled');
+    $("#sending-indicator").show();
+    compose.send_message();
+    assert.equal(local_id_counter_stub, 1);
+    assert.equal(get_events_running_called, 1);
+    assert.equal(reify_message_id_checked, 1);
+    assert.equal(send_msg_ajax_post_called, 1);
+    assert.equal(server_events_triggered, 1);
+    assert.equal(_.keys(compose.send_times_data).length, 1);
+    assert.equal($("#new_message_content").val(), '');
+    assert($("#new_message_content").is_focused());
+    assert(!$("#send-status").visible());
+    assert.equal($("#compose-send-button").attr('disabled'), undefined);
+    assert(!$("#sending-indicator").visible());
+
+    compose_state.set_message_type('stream');
+    var server_error_triggered = false;
+    channel.post = function (payload) {
+        payload.error('500', 'Internal Server Error');
+        send_msg_ajax_post_called = send_msg_ajax_post_called + 1;
+        server_error_triggered = true;
+    };
+    var reload_initiate_triggered = false;
+    global.reload = {
+        is_pending: function () { return true; },
+        initiate: function () {
+            reload_initiate_triggered = true;
+        },
+    };
+    compose.send_message();
+    assert.equal(_.keys(compose.send_times_data).length, 1);
+    assert.equal(local_id_counter_stub, 2);
+    assert.equal(send_msg_ajax_post_called, 2);
```

Comment:
As we discussed over chat, we should use something like a Dict, or maybe a plain JS object is fine, to hold all these various counts, so they're easy to clear.

This test is pretty easy to read, in some sense, since there's not much branching and the mocking is straightforward enough, but just do its sheer length I would put some thought into how make the test feel more approachable.  One possibility is to simply break it up into a couple different tests and have some helper methods.  Or just adding whitespace and a few key comments can go a long way.